src/Pool.sol

```diff
@@ -419,14 +445,21 @@ abstract contract Pool {
             if (receiverAddr == ReceiverWeightsImpl.ADDR_ROOT) break;
             int128 amtPerSecDelta = int128(uint128(weight)) * amtPerWeightPerSecDelta;
             _setReceiverDeltaFromNow(receiverAddr, amtPerSecDelta, timeEnd);
-            if (amtPerSecDelta > 0) {
-                // Sending is starting
-                uint128 amtPerSec = uint128(amtPerSecDelta);
-                emit SenderToReceiverUpdated(id, receiverAddr, amtPerSec, timeEnd);
-            } else {
-                // Sending is stopping
-                emit SenderToReceiverUpdated(id, receiverAddr, 0, _currTimestamp());
+
+            // Find an existing receiver stream log to update
+            uint256 updated = 0;
+            while (updated < updates.length && updates.updates[updated].receiver != receiverAddr) {
+                updated++;
+            }
```

Comment:
I understand the code but I still need to understand why we need to deduplicate the events for subSenders.

I will look at the other PRs to understand it and comeback.

Obviously this while loop is not so nice for the performance.


I think there are some potential optimisations:

```solidity
// lengthBefore is the length before starting to iterate the linked list
 while (updated < lengthBefore && updates.updates[updated].receiver != receiverAddr) {
                updated++;
            }
// New receiver, will be pushed
 if (updated == lengthBefore) {
        updates.length++;
        updated = updates.length
 }
            
            
```
The other one:
```solidity
   // counterHit initialized with zero
   uint256 updated = lastHit;
    while (updated < updates.length && updates.updates[updated].receiver != receiverAddr) {
                updated++;
    }
    if(updated < updates.length) {
        lastHit=updated;
    }
```
I think this should work because the order in the array is very similar to the linked list.

In the startSending case the array will only contain the previous elements and the removed elements. (before iteration of the linked-list)

We don't know the position of the first match from the linked-list in the array but we know the nextMatch needs to come after the first Match (updates.updates[updated].receiver == receiverAddr)).

If we have the first match it means in the startSending case we already passed all newly added elements in the linkedList.

Thats my reasoning here. Maybe I miss something.

Reply:
Great ideas! The first optimization does yield up to a few hundred gas per test case. The second is worsening the results, there's probably too much to manage on the stack and memory access is really cheap. One surprisingly good thing to do in terms of gas is reverting to separate paths for stopping and starting sending, I'm not sure why. Pushed.

Reply:
Interesting.

Reply:
> One surprisingly good thing to do in terms of gas is reverting to separate paths for stopping and starting sending, I'm not sure why.

Mhm, I think previously you had two IFs in the shorthand form now it is just one.

Reply:
Previously we also had the while loop iteration in the stopSending case.

Reply:
> previously you had two IFs in the shorthand form
I've tried both the two ternary operators and the single `if`, both were worse than this.
> Previously we also had the while loop iteration in the stopSending case.
In `stopSending` case it was making 0 iterations, the first check of the first `while` condition check was immediately skipping the loop. To me it should be just as cheap as having an `if` skipping a part of code.